Add third_party module callbacks to check if a user can delete a room and deactivate a user (#12028)
matrix-org/synapse: docs/modules/third_party_rules_callbacks.md

If multiple modules implement this callback, Synapse runs them all in order.

### `check_can_shutdown_room`

_First introduced in Synapse v1.5X.0_

```python
async def check_can_shutdown_room(

>     user_id: str
>     room_id: str,

Reviewer: ```suggestion
    user_id: str, room_id: str
```

) -> bool:
```

Called when an admin user requests the shutdown of a room. The module must return a
boolean indicating whether the shutdown can go through. If the callback returns `False`,

> the shutdown will not proceed and the caller will see a `M_FOBIDDEN` error.

Reviewer 2: ```suggestion
the shutdown will not proceed and the caller will see a `M_FORBIDDEN` error.
```

If multiple modules implement this callback, they will be considered in order. If a
callback returns `True`, Synapse falls through to the next one. The value of the first
callback that does not return `True` will be used. If this happens, Synapse will not call
any of the subsequent implementations of this callback.

### `check_can_deactivate_user`

_First introduced in Synapse v1.5X.0_

```python
async def check_can_deactivate_user(

>     requester: "synapse.types.Requester",

Reviewer: Thinking about this a bit more, is there a reason we need to pass the `Requester` down to the module instead of just the user ID? If not I'd prefer if we passed the user ID for consistency (I know `on_create_room` already passes a `Requester` to modules, we'll probably need to fix it later).

Author: Oh, it was mostly so you could tell if the requester was an admin or not by inspecting the object. I can do an `is_admin` flag (or maybe there is a Module API to determine this) but I was going for convenience.

Author: Removed Requester, second parameter is now either the admin's user id, or None if no admin made the request.

Author: Having thought about it...I can't think of a reason to care who the admin is... going to use `by_admin`

    user_id: str,
) -> bool:
```

Called when the deactivation of a user is requested. User deactivation can be

> performed by an admin or the user themselves, so developers are encouraged to check the

Reviewer: I think there should be a `by_admin` boolean to the signature. This would be consistent with `on_user_deactivation_status_changed`.

Author: Aha, you answered my question.

Author: (cont in the thread above)

requester when implementing this callback. The module must return a
boolean indicating whether the deactivation can go through. If the callback returns `False`,

> the deactivation will not proceed and the caller will see a `M_FOBIDDEN` error.

Reviewer 2: ```suggestion
the deactivation will not proceed and the caller will see a `M_FORBIDDEN` error.
```

If multiple modules implement this callback, Synapse runs them all in order.

### `check_can_shutdown_room`

_First introduced in Synapse v1.5X.0_

```python
async def check_can_shutdown_room(
    user_id: str
    room_id: str,
) -> bool:
```

> Called when a user requests the shutdown of a room. The module must return a

Reviewer 3: ```suggestion
Called when an admin user requests the shutdown of a room. The module must return a
```

I think it's good to spell it out that this affects admin users.

boolean indicating whether the shutdown can go through. If the callback returns `False`,
the shutdown will not proceed and the caller will see a `M_FOBIDDEN` error.

If multiple modules implement this callback, they will be considered in order. If a
callback returns `True`, Synapse falls through to the next one. The value of the first
callback that does not return `True` will be used. If this happens, Synapse will not call
any of the subsequent implementations of this callback.

### `check_can_deactivate_user`

_First introduced in Synapse v1.5X.0_

```python
async def check_can_deactivate_user(
    requester: "synapse.types.Requester",
    user_id: str,
) -> bool:
```

> Called when a user requests the deactivation of a user. User deactivation may be

Reviewer 3: ```suggestion
Called when a user (including admins) requests the deactivation of a user. User deactivation may be
```

Reviewer: This sounds superfluous given the next sentence says:

> User deactivation may be performed by an admin or the user themselves, so developers are encouraged to check the requester when implementing this callback.

Author: what he said. I think this is okay?

Reviewer: Let's say it is.

> Called when a user requests the deactivation of a user. User deactivation may be

Reviewer: ```suggestion
Called when the deactivation of a user is requested. User deactivation can be
```

If multiple modules implement this callback, Synapse runs them all in order.

### `check_can_shutdown_room`

> _First introduced in Synapse v1.5X.0_

Reviewer: ```suggestion
_First introduced in Synapse v1.55.0_
```

```python
async def check_can_shutdown_room(
    user_id: str, room_id: str,
) -> bool:
```

Called when an admin user requests the shutdown of a room. The module must return a
boolean indicating whether the shutdown can go through. If the callback returns `False`,
the shutdown will not proceed and the caller will see a `M_FORBIDDEN` error.

If multiple modules implement this callback, they will be considered in order. If a
callback returns `True`, Synapse falls through to the next one. The value of the first
callback that does not return `True` will be used. If this happens, Synapse will not call
any of the subsequent implementations of this callback.

### `check_can_deactivate_user`

> _First introduced in Synapse v1.5X.0_

Reviewer: ```suggestion
_First introduced in Synapse v1.55.0_
```

```python
async def check_can_deactivate_user(

>     user_id: str,
>     by_admin: bool,

Reviewer: ```suggestion
    user_id: str, by_admin: bool,
```

(to calm down my ocd)

) -> bool:
```

Called when the deactivation of a user is requested. User deactivation can be
performed by an admin or the user themselves, so developers are encouraged to check the
requester when implementing this callback. The module must return a
boolean indicating whether the deactivation can go through. If the callback returns `False`,
the deactivation will not proceed and the caller will see a `M_FORBIDDEN` error.

> `by_admin` will be `True` if the request is made by an admin, or `False` if the request
> was made by the user themselves.

Reviewer: ```suggestion
The module is passed two parameters, `user_id` which is the ID of the user being deactivated, and `by_admin` which is `True` if the request is made by a serve admin, and `False` otherwise.
```

Bit of clarification. Also, `True` doesn't necessarily mean that the deactivation wasn't made by the user themselves (it will be `True` if a server admin tries to deactivate their own account using the admin API).

Author: Good points.
